File: tool/sync_musify.dart
#!/usr/bin/env dart

import 'dart:io';
import 'dart:convert';

const String upsteamRepo = 'https://github.com/gokadzev/Musify.git';
const String upstreamBranch = 'master';
const String yourBranch = 'harmonyos-port';

/// 颜色输出
class Colors {
  static const String reset = '\x1B[0m';
  static const String red = '\x1B[31m';
  static const String green = '\x1B[32m';
  static const String yellow = '\x1B[33m';
  static const String blue = '\x1B[34m';
  static const String cyan = '\x1B[36m';

  static void info(String msg) => print('$blue[INFO]$reset $msg');
  static void success(String msg) => print('$green[SUCCESS]$reset $msg');
  static void warning(String msg) => print('$yellow[WARNING]$reset $msg');
  static void error(String msg) => print('$red[ERROR]$reset $msg');
  static void title(String msg) => print('\n$cyan=== $msg ===$reset\n');
}

/// 执行命令并返回结果
Future<ProcessResult> runCommand(String cmd, List<String> args,
    {String? workingDirectory}) async {
  return await Process.run(cmd, args,
      workingDirectory: workingDirectory, runInShell: true);
}

/// 检查是否在 Git 仓库中
Future<bool> isGitRepo() async {
  final result = await runCommand('git', ['rev-parse', '--git-dir']);
  return result.exitCode == 0;
}

/// 检查工作区是否干净
Future<bool> isWorkspaceClean() async {
  final result = await runCommand('git', ['diff-index', '--quiet', 'HEAD', '--']);
  return result.exitCode == 0;
}

/// 获取未提交的文件列表
Future<List<String>> getUncommittedFiles() async {
  final result = await runCommand('git', ['status', '--porcelain']);
  if (result.exitCode != 0) return [];
  return (result.stdout as String)
      .split('\n')
      .where((line) => line.trim().isNotEmpty)
      .toList();
}

/// 添加 upstream 远程仓库
Future<void> setupUpstream() async {
  final result = await runCommand('git', ['remote']);
  final remotes = (result.stdout as String).split('\n');

  if (remotes.contains('upstream')) {
    Colors.info('upstream 已存在，更新 URL...');
    await runCommand('git', ['remote', 'set-url', 'upstream', upsteamRepo]);
  } else {
    Colors.info('添加 upstream 远程仓库...');
    await runCommand('git', ['remote', 'add', 'upstream', upsteamRepo]);
  }
  Colors.success('upstream 配置完成');
}

/// 创建备份分支
Future<String> createBackup() async {
  final timestamp = DateTime.now().toIso8601String().replaceAll(':', '-').split('.')[0];
  final backupBranch = 'backup-$timestamp';
  Colors.info('创建备份分支: $backupBranch');
  await runCommand('git', ['branch', backupBranch]);
  Colors.success('备份分支创建成功');
  return backupBranch;
}

/// 获取上游更新日志
Future<void> showUpdateLog() async {
  final result = await runCommand(
      'git', ['log', 'HEAD..upstream/$upstreamBranch', '--oneline', '--no-decorate']);
  if (result.exitCode == 0 && (result.stdout as String).trim().isNotEmpty) {
    Colors.info('即将合并的更新：');
    print(result.stdout);
  } else {
    Colors.info('没有新更新');
  }
}

/// 执行合并
Future<bool> performMerge() async {
  Colors.info('开始合并 upstream/$upstreamBranch ...');

  final mergeResult = await runCommand(
      'git', ['merge', 'upstream/$upstreamBranch', '--no-edit']);

  if (mergeResult.exitCode == 0) {
    Colors.success('合并成功！');
    return true;
  } else {
    Colors.warning('合并出现冲突');
    print(mergeResult.stderr);

    // 获取冲突文件列表
    final conflictResult = await runCommand('git', ['diff', '--name-only', '--diff-filter=U']);
    if (conflictResult.exitCode == 0) {
      final conflicts = (conflictResult.stdout as String)
          .split('\n')
          .where((f) => f.trim().isNotEmpty)
          .toList();

      if (conflicts.isNotEmpty) {
        Colors.info('冲突文件列表：');
        for (final file in conflicts) {
          print('  ⚠️  $file');
        }
      }
    }

    print('\n${Colors.yellow}建议处理方式：');
    print('  1. 查看冲突文件：git status');
    print('  2. 手动编辑冲突文件，删除 <<<<<<< ======= >>>>>>> 标记');
    print('  3. 对于鸿蒙特定文件，保留你的修改版本');
    print('  4. 解决完冲突后执行：git add . && git commit');
    print('\n  或者使用自动解决（谨慎）：');
    print('  git merge --abort && git merge -X ours upstream/$upstreamBranch');

    return false;
  }
}

/// 运行 Flutter 依赖更新
Future<void> runFlutterPubGet() async {
  final flutterResult = await runCommand('flutter', ['--version']);
  if (flutterResult.exitCode == 0) {
    Colors.info('运行 flutter pub get 更新依赖...');
    await runCommand('flutter', ['pub', 'get']);
    Colors.success('依赖更新完成');
  } else {
    Colors.warning('Flutter 未找到，请手动运行 flutter pub get');
  }
}

/// 检查并切换到鸿蒙分支
Future<bool> ensureHarmonyBranch() async {
  // 获取当前分支
  final branchResult = await runCommand('git', ['branch', '--show-current']);
  final currentBranch = (branchResult.stdout as String).trim();

  if (currentBranch == yourBranch) {
    Colors.info('已在 $yourBranch 分支');
    return true;
  }

  // 检查分支是否存在
  final branchList = await runCommand('git', ['branch']);
  final branches = branchList.stdout as String;

  if (branches.contains(yourBranch)) {
    Colors.info('切换到分支: $yourBranch');
    await runCommand('git', ['checkout', yourBranch]);
    return true;
  } else {
    Colors.error('分支 $yourBranch 不存在');
    Colors.info('可用分支：');
    print(branches);
    return false;
  }
}

/// 主同步流程
Future<void> syncMain() async {
  Colors.title('开始同步 Musify 更新');

  // 1. 拉取上游代码
  Colors.info('拉取 upstream 最新代码...');
  final fetchResult = await runCommand('git', ['fetch', 'upstream']);
  if (fetchResult.exitCode != 0) {
    Colors.error('拉取失败');
    print(fetchResult.stderr);
    exit(1);
  }

  // 2. 显示更新日志
  await showUpdateLog();

  // 3. 询问是否继续
  stdout.write('\n是否继续合并？(y/N) ');
  final input = await stdin.first;
  if (input.toString().toLowerCase() != 'y\n') {
    Colors.warning('用户取消同步');
    return;
  }

  // 4. 执行合并
  final success = await performMerge();
  if (!success) {
    exit(1);
  }

  // 5. 更新依赖
  await runFlutterPubGet();

  Colors.title('同步完成');
}

/// 主函数
Future<void> main(List<String> args) async {
  // 解析命令行参数
  if (args.contains('-h') || args.contains('--help')) {
    print('''
Musify 鸿蒙版本同步脚本 (Dart)

用法: dart tool/sync_musify.dart [选项]

选项:
    -h, --help      显示此帮助信息
    -c, --check     仅检查仓库状态，不同步
    -b, --backup    创建备份分支后退出

示例:
    dart tool/sync_musify.dart           执行同步
    dart tool/sync_musify.dart --check   检查状态
    dart tool/sync_musify.dart --backup  创建备份
''');
    return;
  }

  if (args.contains('-c') || args.contains('--check')) {
    if (!await isGitRepo()) {
      Colors.error('当前目录不是 Git 仓库');
      exit(1);
    }
    Colors.info('当前仓库状态：');
    await runCommand('git', ['remote', '-v']);
    print('');
    await runCommand('git', ['branch', '-a']);
    print('');
    await runCommand('git', ['status']);
    return;
  }

  if (args.contains('-b') || args.contains('--backup')) {
    if (!await isGitRepo()) {
      Colors.error('当前目录不是 Git 仓库');
      exit(1);
    }
    await createBackup();
    Colors.success('备份创建完成');
    return;
  }

  // 主流程
  if (!await isGitRepo()) {
    Colors.error('当前目录不是 Git 仓库');
    exit(1);
  }

  if (!await isWorkspaceClean()) {
    final files = await getUncommittedFiles();
    Colors.warning('工作区有未提交的修改：');
    for (final file in files) {
      print('  📝 $file');
    }
    stdout.write('\n是否继续？(y/N) ');
    final input = await stdin.first;
    if (input.toString().toLowerCase() != 'y\n') {
      Colors.info('请先提交或暂存你的修改');
      exit(1);
    }
  }

  await setupUpstream();
  await createBackup();

  if (!await ensureHarmonyBranch()) {
    exit(1);
  }

  await syncMain();
}

File: tool/sync-musify.ts
#!/usr/bin/env bun

import { $, spawn } from 'bun';

// 配置
const CONFIG = {
  upstreamRepo: 'https://github.com/gokadzev/Musify.git',
  upstreamBranch: 'master',
  yourBranch: 'harmonyos-port',
};

// 颜色输出
const c = {
  reset: '\x1B[0m',
  red: '\x1B[31m',
  green: '\x1B[32m',
  yellow: '\x1B[33m',
  blue: '\x1B[34m',
  cyan: '\x1B[36m',

  info: (msg: string) => console.log(`${c.blue}[INFO]${c.reset} ${msg}`),
  success: (msg: string) => console.log(`${c.green}[SUCCESS]${c.reset} ${msg}`),
  warning: (msg: string) => console.log(`${c.yellow}[WARNING]${c.reset} ${msg}`),
  error: (msg: string) => console.log(`${c.red}[ERROR]${c.reset} ${msg}`),
  title: (msg: string) => console.log(`\n${c.cyan}=== ${msg} ===${c.reset}\n`),
};

// 执行命令并返回结果
async function run(cmd: string[], options?: { cwd?: string }): Promise<{ success: boolean; stdout: string; stderr: string }> {
  try {
    const result = await $`${cmd}`.cwd(options?.cwd ?? process.cwd()).quiet();
    return { success: true, stdout: result.text(), stderr: '' };
  } catch (error: any) {
    return { success: false, stdout: error.stdout?.toString() ?? '', stderr: error.stderr?.toString() ?? '' };
  }
}

// 检查是否在 Git 仓库中
async function isGitRepo(): Promise<boolean> {
  const result = await run(['git', 'rev-parse', '--git-dir']);
  return result.success;
}

// 检查工作区是否干净
async function isWorkspaceClean(): Promise<boolean> {
  const result = await run(['git', 'diff-index', '--quiet', 'HEAD', '--']);
  return result.success;
}

// 获取未提交的文件
async function getUncommittedFiles(): Promise<string[]> {
  const result = await run(['git', 'status', '--porcelain']);
  if (!result.success) return [];
  return result.stdout.split('\n').filter(l => l.trim().length > 0);
}

// 设置 upstream
async function setupUpstream(): Promise<void> {
  const result = await run(['git', 'remote']);
  const remotes = result.stdout.split('\n');

  if (remotes.includes('upstream')) {
    c.info('upstream 已存在，更新 URL...');
    await run(['git', 'remote', 'set-url', 'upstream', CONFIG.upstreamRepo]);
  } else {
    c.info('添加 upstream 远程仓库...');
    await run(['git', 'remote', 'add', 'upstream', CONFIG.upstreamRepo]);
  }
  c.success('upstream 配置完成');
}

// 创建备份分支
async function createBackup(): Promise<string> {
  const timestamp = new Date().toISOString().replace(/:/g, '-').split('.')[0];
  const backupBranch = `backup-${timestamp}`;
  c.info(`创建备份分支: ${backupBranch}`);
  await run(['git', 'branch', backupBranch]);
  c.success('备份分支创建成功');
  return backupBranch;
}

// 显示更新日志
async function showUpdateLog(): Promise<void> {
  const result = await run(['git', 'log', `HEAD..upstream/${CONFIG.upstreamBranch}`, '--oneline', '--no-decorate']);
  if (result.success && result.stdout.trim()) {
    c.info('即将合并的更新：');
    console.log(result.stdout);
  } else {
    c.info('没有新更新');
  }
}

// 执行合并
async function performMerge(): Promise<boolean> {
  c.info(`开始合并 upstream/${CONFIG.upstreamBranch} ...`);

  const mergeResult = await run(['git', 'merge', `upstream/${CONFIG.upstreamBranch}`, '--no-edit']);

  if (mergeResult.success) {
    c.success('合并成功！');
    return true;
  } else {
    c.warning('合并出现冲突');
    console.log(mergeResult.stderr);

    // 获取冲突文件
    const conflictResult = await run(['git', 'diff', '--name-only', '--diff-filter=U']);
    if (conflictResult.success && conflictResult.stdout.trim()) {
      const conflicts = conflictResult.stdout.split('\n').filter(f => f.trim());
      c.info('冲突文件列表：');
      conflicts.forEach(f => console.log(`  ⚠️  ${f}`));
    }

    console.log(`\n${c.yellow}建议处理方式：${c.reset}`);
    console.log('  1. 查看冲突文件：git status');
    console.log('  2. 手动编辑冲突文件，删除 <<<<<<< ======= >>>>>>> 标记');
    console.log('  3. 对于鸿蒙特定文件，保留你的修改版本');
    console.log('  4. 解决完冲突后执行：git add . && git commit');

    return false;
  }
}

// 更新 Flutter 依赖
async function runFlutterPubGet(): Promise<void> {
  const flutterCheck = await run(['flutter', '--version']);
  if (flutterCheck.success) {
    c.info('运行 flutter pub get 更新依赖...');
    await run(['flutter', 'pub', 'get']);
    c.success('依赖更新完成');
  } else {
    c.warning('Flutter 未找到，请手动运行 flutter pub get');
  }
}

// 切换到鸿蒙分支
async function ensureHarmonyBranch(): Promise<boolean> {
  const branchResult = await run(['git', 'branch', '--show-current']);
  const currentBranch = branchResult.stdout.trim();

  if (currentBranch === CONFIG.yourBranch) {
    c.info(`已在 ${CONFIG.yourBranch} 分支`);
    return true;
  }

  const branchList = await run(['git', 'branch']);
  if (branchList.stdout.includes(CONFIG.yourBranch)) {
    c.info(`切换到分支: ${CONFIG.yourBranch}`);
    await run(['git', 'checkout', CONFIG.yourBranch]);
    return true;
  } else {
    c.error(`分支 ${CONFIG.yourBranch} 不存在`);
    console.log(branchList.stdout);
    return false;
  }
}

// 主同步流程
async function syncMain(): Promise<void> {
  c.title('开始同步 Musify 更新');

  // 拉取上游
  c.info('拉取 upstream 最新代码...');
  const fetchResult = await run(['git', 'fetch', 'upstream']);
  if (!fetchResult.success) {
    c.error('拉取失败');
    console.log(fetchResult.stderr);
    process.exit(1);
  }

  // 显示更新
  await showUpdateLog();

  // 询问确认
  const encoder = new TextEncoder();
  const input = await (async () => {
    process.stdout.write('\n是否继续合并？(y/N) ');
    const buffer = new Uint8Array(1);
    await Bun.stdin.read(buffer);
    return String.fromCharCode(buffer[0]);
  })();

  if (input.toLowerCase() !== 'y') {
    c.warning('用户取消同步');
    return;
  }

  // 执行合并
  const success = await performMerge();
  if (!success) process.exit(1);

  // 更新依赖
  await runFlutterPubGet();

  c.title('同步完成');
}

// 显示帮助
function showHelp(): void {
  console.log(`
Musify 鸿蒙版本同步脚本 (Bun)

用法: bun run sync-musify.ts [选项]

选项:
    -h, --help      显示此帮助信息
    -c, --check     仅检查仓库状态，不同步
    -b, --backup    创建备份分支后退出

示例:
    bun run sync-musify.ts           执行同步
    bun run sync-musify.ts --check   检查状态
    bun run sync-musify.ts --backup  创建备份
`);
}

// 主函数
async function main() {
  const args = process.argv.slice(2);

  if (args.includes('-h') || args.includes('--help')) {
    showHelp();
    return;
  }

  if (args.includes('-c') || args.includes('--check')) {
    if (!await isGitRepo()) {
      c.error('当前目录不是 Git 仓库');
      process.exit(1);
    }
    c.info('当前仓库状态：');
    await run(['git', 'remote', '-v']);
    console.log('');
    await run(['git', 'branch', '-a']);
    console.log('');
    await run(['git', 'status']);
    return;
  }

  if (args.includes('-b') || args.includes('--backup')) {
    if (!await isGitRepo()) {
      c.error('当前目录不是 Git 仓库');
      process.exit(1);
    }
    await createBackup();
    c.success('备份创建完成');
    return;
  }

  // 主流程
  if (!await isGitRepo()) {
    c.error('当前目录不是 Git 仓库');
    process.exit(1);
  }

  if (!await isWorkspaceClean()) {
    const files = await getUncommittedFiles();
    c.warning(`工作区有 ${files.length} 个未提交的修改`);
    files.forEach(f => console.log(`  📝 ${f}`));

    process.stdout.write('\n是否继续？(y/N) ');
    const encoder = new TextEncoder();
    const buffer = new Uint8Array(1);
    await Bun.stdin.read(buffer);
    const input = String.fromCharCode(buffer[0]);

    if (input.toLowerCase() !== 'y') {
      c.info('请先提交或暂存你的修改');
      process.exit(1);
    }
  }

  await setupUpstream();
  await createBackup();

  if (!await ensureHarmonyBranch()) process.exit(1);

  await syncMain();
}

// 运行
main().catch(console.error);